# Pandas
1. Pandas
2. Data structures
3. Pandas Series Series
     - Creating Series 
     - Manipulating Series
4. Pandas Dataframes
    - Creating Dataframes
    - Manipulating Dataframes
5. Reading Data from different Sources

## 1. Pandas
- It contains data structures and manipulation tools designed for data cleaning and analysis. 
- While it adopts many idioms from _Numpy_, there biggest difference is that Pandas is desined for working with tabular ot hereogenious data.NumPy , by contrast , is best suited for working with homogeneous numerical array data.
- Its name is derieved from "Panel data" an econometrics term for multidimensional structured data sets.

### Pandas installation and import 
- installation 
`!pip install pandas`
- Import 
`import pandas as pd` 

In [1]:
# Import Pandas 
import pandas as pd

## 2. Data Structures
- Pandas has 2 data structures as follows:
1. A __Series__ is 1-dimensional labeled array that can hold data of any type (integer, float,string ,boolean,python object,and so on). Its axis labels are collectively called an index. 
2. A __DataFrame__ is a 2- dimensional labelled data structure with columns. it supports multiple datatypes.

## 3. Pandas Series 
- Is a one- dimensional labeled arrau capable of holding any data type. However,a series is a sequence of homogenoues data types, similar to an array , list , or column in a tabe.
- It will assign a labeled index to each item in the Series. By default, each item will receive an index label from 0 to N, where N is the length of the Series minus one.

### 3.1 Creating Series 
1. __To create a numeric series__.

In [4]:
# Create a numeric series  
numbers = range(1,100,5)
pd.Series(numbers)

0      1
1      6
2     11
3     16
4     21
5     26
6     31
7     36
8     41
9     46
10    51
11    56
12    61
13    66
14    71
15    76
16    81
17    86
18    91
19    96
dtype: int64

- The output is of type `int64`
- The row names are usually denoted as _"Index"_
2. __To create an object series__

In [5]:
string = "Hi" , "How " , "are " ,"you","?"
pd.Series(string)

0      Hi
1    How 
2    are 
3     you
4       ?
dtype: object

- Output is of type `object`
3. __To create a series with both numeric and string values__

In [6]:
# create a series with an arbituary list 
pd.Series([365,'London',34.5,-34.5,'Happy Birthday'])

0               365
1            London
2              34.5
3             -34.5
4    Happy Birthday
dtype: object

- Here numeric types are treated as objects. _A serie cannot have multiple data types so it defines all of them as an object_
4. To set index values for a series.

In [7]:
marks = [60,89,74,86,100]

subject = ["Math" , "System design" , "Cloud Computing" , "Data Analysis" , "React"]

marks_series= pd.Series(marks,index =subject)
marks_series


Math                60
System design       89
Cloud Computing     74
Data Analysis       86
React              100
dtype: int64

- the index is added usin the argument `index=`. The data tyoe if the series continues to be numeric.

6. __To create series from a dictionary__

In [8]:
data = {'React':90,"Node":85,"Flutter":50,"Django":75}
pd.Series(data)

React      90
Node       85
Flutter    50
Django     75
dtype: int64

- On passing a dict the index in the resulting Series will have the dict's keys in sorted oder.

6. __A series with missing values__
- If we pass a key that is not defined then its value will be `NAN`

In [9]:
subject = ["Math" , "System design" , "Cloud Computing" , "Data Analysis" , "React"] 

marks_series = pd.Series(data,index=subject)
marks_series

Math                NaN
System design       NaN
Cloud Computing     NaN
Data Analysis       NaN
React              90.0
dtype: float64

In [10]:
# Error
index=['Apple', 'Banana', 'Orange']
quantity = [34, 20, 30, 40]
# Uncomment to see error 👇🏿.
# pd.Series(data=quantity, index=index)

In [11]:
dict={'A':30, 'B':40, 'C':50}
index=['A', 'B', 'D']
pd.Series(data=dict, index=index)

A    30.0
B    40.0
D     NaN
dtype: float64

### 3.2 Manipulating Series
1. __To check for null values using `.notnull`__

In [12]:
marks_series.notnull()

Math               False
System design      False
Cloud Computing    False
Data Analysis      False
React               True
dtype: bool

- `True` indicates that the value is not null.
3. __To know the subjects in which marks score is more than 75__

In [13]:
marks_series[marks_series > 75]

React    90.0
dtype: float64

4. ___To assign 68 marks to 'Art and Craft'__

In [16]:
marks_series["Math"] = 91
# or
#  mark
marks_series

Math               91.0
System design       NaN
Cloud Computing     NaN
Data Analysis       NaN
React              91.0
dtype: float64

In [17]:
# Compare values
marks_series["Math"] == 75
# OR
marks_series.Math == 75

False

5. __Sorting Numeric Series__

In [3]:
import numpy as np

In [19]:
values = pd.Series([23,np.nan,45,np.nan,56,67,34,23])
values

0    23.0
1     NaN
2    45.0
3     NaN
4    56.0
5    67.0
6    34.0
7    23.0
dtype: float64

In [20]:
#ascending Order
values.sort_values(ascending = True)
#descending Order
values.sort_values(ascending = False)

5    67.0
4    56.0
2    45.0
6    34.0
0    23.0
7    23.0
1     NaN
3     NaN
dtype: float64

7. __Sorting Categorical Series__

In [21]:
# create a pandas series 
string_values = pd.Series(["a", "f", "j", "d", "c"])
string_values
# since the computer stores strings in lexigraphical order
# sort_values maintains the indices of all the elements of the array

0    a
1    f
2    j
3    d
4    c
dtype: object

In [22]:
# ascending order
string_values.sort_values(ascending=True)

0    a
4    c
3    d
1    f
2    j
dtype: object

In [23]:
data=range(10)
new_ser=pd.Series(data=data)
new_ser[new_ser==5]

5    5
dtype: int64

In [24]:
marks_series.rank(ascending=True)

Math               1.5
System design      NaN
Cloud Computing    NaN
Data Analysis      NaN
React              1.5
dtype: float64

# Panda Data Frames
- Is a tabular representation of data containing an ordered collectin, each of which can ve a different type (numeric , string,boolean and so on)
- The DataFrame has both row and column index; it can be thought of as a dict of Series all sharing the same index.In a DF, the data is stored as one or more two-dimensioanl blocks rather than a list,dict or some other collection of one-dimensional arrays.
- While a DF is physically two-dimensional, it can be use to represent higher dimensional data in  tabular   format usung hierarchical indexing

- __4.1 Creating a DataFrame__

1. __Creating a DataFrame from a dictionary__

In [25]:
data ={
    'Subject':["React","Rust","Golang","Elixir"],
    'Marks':(89,34,65,78), 
    'CGPA':[2.5,3.0,4.5,5.6]
}
df = pd.DataFrame(data)
df

Subject  Marks  CGPA
0   React     89   2.5
1    Rust     34   3.0
2  Golang     65   4.5
3  Elixir     78   5.6

__Note__: Like Series, the resulting DataFrame is assigned index automatically. ANd the "Marks" values are in a tuple.

2. __To Create a DataFrame From series__

In [3]:
Subject = pd.Series(["Math","DSA","Python","ML"])
Marks = pd.Series([23,23,23,23])
CGPA = pd.Series([2.5,3.0,4.0,6.0])

In [4]:
pd.DataFrame([Subject,Marks,CGPA],index = ['Subject','Marks','CGPA'])

0    1       2    3
Subject  Math  DSA  Python   ML
Marks      23   23      23   23
CGPA      2.5  3.0     4.0  6.0

- However to want a vertical data frame we use `.T` . The 'T' stands for _transpose_.

In [5]:
pd.DataFrame([Subject,Marks,CGPA],index = ['Subject','Marks','CGPA']).T

Subject Marks CGPA
0    Math    23  2.5
1     DSA    23  3.0
2  Python    23  4.0
3      ML    23  6.0

4. __To create a dataframe from lists__

In [6]:
Subject = ["Math","DSA","Python","ML"]
Marks = [23,23,23,23]
CGPA = [2.5,3.0,4.0,6.0]

pd.DataFrame([Subject,Marks,CGPA],index=['Subject','Marks','CGPA'])

0    1       2    3
Subject  Math  DSA  Python   ML
Marks      23   23      23   23
CGPA      2.5  3.0     4.0  6.0

In [7]:
pd.DataFrame([Subject,Marks,CGPA],index=['Subject','Marks','CGPA']).T

Subject Marks CGPA
0    Math    23  2.5
1     DSA    23  3.0
2  Python    23  4.0
3      ML    23  6.0

In [18]:
a1 = ['Hogwarts', 'Durmstrang', 'Beauxbatons']
a2 = ['Hogwarts', 'Durmstrang', 'Beauxbatons']
a3 = ['Hogwarts', 'Durmstrang', 'Beauxbatons']
school = [a1, a2, a3]
inst = ['School_1', 'School_2', 'School_3']
Muggle_data = pd.DataFrame(data=school, columns=inst)
Muggle_data

School_1    School_2     School_3
0  Hogwarts  Durmstrang  Beauxbatons
1  Hogwarts  Durmstrang  Beauxbatons
2  Hogwarts  Durmstrang  Beauxbatons

5. __To read data from csv file__


In [6]:
data = pd.read_csv("sample.csv")
type(data)
# The file must be in the same directory as the notebook.

pandas.core.frame.DataFrame

- On checking the data type it is a pandas Dataframe.
6. __To print the head of the data__

In [10]:
data.head() # 1st 5 elements

name   world_ranking region_ranking  \
0                   Aarhus University  #150 of 14,131   #44 of 2,785   
1    Arizona State University - Tempe   #61 of 14,131   #48 of 2,597   
2                   Auburn University             NaN            NaN   
3      Australian National University   #88 of 14,131       #5 of 59   
4  Autonomous University of Barcelona             NaN            NaN   

  country_ranking         region        country      city\state  \
0        #2 of 27         Europe          Italy          Veneto   
1    #45 of 2,496         Europe          Spain        Valencia   
2             NaN  North America  United States       Wisconsin   
3        #5 of 40  North America         Canada         Ontario   
4             NaN         Europe          Italy  Emilia-Romagna   

  acceptance_rate publication         website            phone_no  \
0             15%      89,633       www.au.dk       +45 8942 1111   
1             NaN      99,086     www.asu.edu          8552785080   
2             57%      36,231      auburn.edu          3348444000   
3             86%      97,754  www.anu.edu.au  +61 (0)2 6125 5111   
4             11%      74,922     www.uab.cat       +34 935812222   

                                             address  
0  Nordre Ringgade 1\n Aarhus, Central Denmark Re...  
1  University Drive and Mill Avenue\n Tempe, Ariz...  
2  Samford Hall\n Auburn, Alabama, 36849 \nUnited...  
3  Ellery Crescent, Acton\n Canberra, Australian ...  
4  Campus de Bellaterra, Edificio A\n Cerdanyola ...

- By deafult, the `.head()` will display first five rows.However we can set the desired number of rows to be displayed. 
`data.head(9)`

7. __To print tail data__


In [11]:
data.tail() # Last 5

name   world_ranking region_ranking country_ranking  \
295  Xi'an Jiaotong University             NaN            NaN             NaN   
296            Yale University   #11 of 14,131    #9 of 2,597     #9 of 2,496   
297          Yonsei University             NaN            NaN             NaN   
298            York University             NaN            NaN             NaN   
299        Zhejiang University  #109 of 14,131    #7 of 5,830       #3 of 960   

            region        country   city\state acceptance_rate publication  \
295  North America  United States  Connecticut              7%      96,097   
296           Asia          China        Xi'an             NaN     198,095   
297  North America  United States   Washington             56%      95,497   
298        Oceania      Australia     Canberra             33%      41,257   
299  North America  United States      Alabama             52%     176,136   

              website             phone_no  \
295   www.xjtu.edu.cn    +86 (29) 266 8830   
296      www.yale.edu           2034324771   
297  www.yonsei.ac.kr    +82 (2) 2123 2114   
298          yorku.ca    +1 (416) 736 5002   
299    www.zju.edu.cn  +86 (571) 8795 1020   

                                               address  
295  28 Xianning Road\n Xi'an, Shaanxi, 710049 \nChina  
296  Woodbridge Hall\n New Haven, Connecticut, 0652...  
297  134 Sinchon-dong, Seodaemun-gu\n Seoul, Seoul,...  
298  4700 Keele Street\n Toronto, Ontario, M3J 1P3 ...  
299  38 Zheda Road, Xihu\n Hangzhou, Zhejiang, 3100...

-  To obtain the dimensions of the data

In [12]:
data.shape
# (rows,columns)

(300, 12)

9. __To know the data types of the data frame__

In [13]:
data.dtypes
# Shows Data type of each variable

name               object
world_ranking      object
region_ranking     object
country_ranking    object
region             object
country            object
city\state         object
acceptance_rate    object
publication        object
website            object
phone_no           object
address            object
dtype: object

10. To know some information of the data

In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   name             300 non-null    object
 1   world_ranking    199 non-null    object
 2   region_ranking   199 non-null    object
 3   country_ranking  199 non-null    object
 4   region           300 non-null    object
 5   country          300 non-null    object
 6   city\state       300 non-null    object
 7   acceptance_rate  224 non-null    object
 8   publication      300 non-null    object
 9   website          294 non-null    object
 10  phone_no         296 non-null    object
 11  address          300 non-null    object
dtypes: object(12)
memory usage: 28.2+ KB


- We see this output gives the number of rows present in the data `RangeIndex: 300 entries, 0 to 299` There are 23 rows numbered from 0 to 22. And there are a total of three columns - `Data columns (total 12 columns):`

Consider `0 name 300 non-null object` indicates that the column named 'Age' has 300 non-null observations having a datatype of `object` and finally the memory used to save this data frame is `memory usage: 28.2+ KB`

11. __To check the data type of column in the data frame__

In [15]:
type(data.name)

pandas.core.series.Series

In [16]:
type(data["publication"])

pandas.core.series.Series

- Note that every column on the data frame is a panda Series

# 4.2 Manipulating DataFrames 

1. __Adding new column to data set__

In [ ]:
data["BMI"] = data["weight"] / data["height"]**2

2. __Add a new row in a data set__
- A new row can be added using the `.copy()`.

In [ ]:
# In the case of tuples , when index is changed .. it changes the ordering of original data frames.
# create a copy of the dataframe
data_copy = data.copy()
# Add a new row at index -> 301 and add infomation in the 
data_copy.loc[301] = ["Zhejiang University"	]

- We see that the new colmn number 301 has been added to the data.

3. __Indexing a dataframe using `.iloc()`__
- `DataFrame.iloc[]` method is used when the index label of the data frame is something other than numeric series of 0,1,2,3...n or incase the user doesnt know the index label.

we shall work on the BMI data-set

__Select the second row__

In [8]:
data.iloc[2] # Selecting the 2nd row

name                                               Auburn University
world_ranking                                                    NaN
region_ranking                                                   NaN
country_ranking                                                  NaN
region                                                 North America
country                                                United States
city\state                                                 Wisconsin
acceptance_rate                                                  57%
publication                                                   36,231
website                                                   auburn.edu
phone_no                                                  3348444000
address            Samford Hall\n Auburn, Alabama, 36849 \nUnited...
Name: 2, dtype: object

__Select 4th,7th and 10th rows__

In [11]:
# Creates a sub data frame from the list passed to it -> sub dataframe with only two rows 7 and 4
data.iloc[[4,7]]

name  world_ranking region_ranking  \
4  Autonomous University of Barcelona            NaN            NaN   
7                   Boston University  #49 of 14,131   #42 of 2,597   

  country_ranking  region country      city\state acceptance_rate publication  \
4             NaN  Europe   Italy  Emilia-Romagna             11%      74,922   
7    #39 of 2,496  Europe  Greece          Athens             NaN     107,676   

       website       phone_no  \
4  www.uab.cat  +34 935812222   
7       bu.edu     6173532000   

                                             address  
4  Campus de Bellaterra, Edificio A\n Cerdanyola ...  
7  One Silber Way\n Boston, Massachusetts, 02215 ...

- We use square brakets since we are passing a list of row numbers to be accessed.

__Select 2nd to 5th rows__

In [12]:
data.iloc[2:5]

name  world_ranking region_ranking  \
2                   Auburn University            NaN            NaN   
3      Australian National University  #88 of 14,131       #5 of 59   
4  Autonomous University of Barcelona            NaN            NaN   

  country_ranking         region        country      city\state  \
2             NaN  North America  United States       Wisconsin   
3        #5 of 40  North America         Canada         Ontario   
4             NaN         Europe          Italy  Emilia-Romagna   

  acceptance_rate publication         website            phone_no  \
2             57%      36,231      auburn.edu          3348444000   
3             86%      97,754  www.anu.edu.au  +61 (0)2 6125 5111   
4             11%      74,922     www.uab.cat       +34 935812222   

                                             address  
2  Samford Hall\n Auburn, Alabama, 36849 \nUnited...  
3  Ellery Crescent, Acton\n Canberra, Australian ...  
4  Campus de Bellaterra, Edificio A\n Cerdanyola ...

- __Select First Column__

In [15]:
# going after the first column
data.iloc[:,0]

0                       Aarhus University
1        Arizona State University - Tempe
2                       Auburn University
3          Australian National University
4      Autonomous University of Barcelona
                      ...                
295             Xi'an Jiaotong University
296                       Yale University
297                     Yonsei University
298                       York University
299                   Zhejiang University
Name: name, Length: 300, dtype: object

In [16]:
# Select last column
data.iloc[:,-1]

0      Nordre Ringgade 1\n Aarhus, Central Denmark Re...
1      University Drive and Mill Avenue\n Tempe, Ariz...
2      Samford Hall\n Auburn, Alabama, 36849 \nUnited...
3      Ellery Crescent, Acton\n Canberra, Australian ...
4      Campus de Bellaterra, Edificio A\n Cerdanyola ...
                             ...                        
295    28 Xianning Road\n Xi'an, Shaanxi, 710049 \nChina
296    Woodbridge Hall\n New Haven, Connecticut, 0652...
297    134 Sinchon-dong, Seodaemun-gu\n Seoul, Seoul,...
298    4700 Keele Street\n Toronto, Ontario, M3J 1P3 ...
299    38 Zheda Road, Xihu\n Hangzhou, Zhejiang, 3100...
Name: address, Length: 300, dtype: object

In [17]:
# Select the first 2  columns
data.iloc[:,0:1]

name
0                     Aarhus University
1      Arizona State University - Tempe
2                     Auburn University
3        Australian National University
4    Autonomous University of Barcelona
..                                  ...
295           Xi'an Jiaotong University
296                     Yale University
297                   Yonsei University
298                     York University
299                 Zhejiang University

[300 rows x 1 columns]

__Selecting the first 2 columns and first 5 cells__

In [18]:
data.iloc[5:7,0:2]

name world_ranking
5  Baylor College of Medicine           NaN
6              Boston College           NaN

4. __Indexing a dataframe using `.loc`__
- `DataFrame.loc[]` method that takes only index labels and returns row or dataframe if the index label exists in the caller data frame.
- `DataFrame.loc[Row_names,column_names]` is used to select or index rows or columns based on their name.

__Select 5 rows and 2nd and 4th coluns__

In [7]:
data.loc[1:5,["name","world_ranking"]]

name  world_ranking
1    Arizona State University - Tempe  #61 of 14,131
2                   Auburn University            NaN
3      Australian National University  #88 of 14,131
4  Autonomous University of Barcelona            NaN
5          Baylor College of Medicine            NaN

- __Note__ the row names are numbers 
5. __Selecting Columns by specifying column names__  
__Select the column `country`__

In [8]:
data.country

0              Italy
1              Spain
2      United States
3             Canada
4              Italy
           ...      
295    United States
296            China
297    United States
298        Australia
299    United States
Name: country, Length: 300, dtype: object

- __Remark__ Using this method we can selsect only one column

In [9]:
data["country"]

0              Italy
1              Spain
2      United States
3             Canada
4              Italy
           ...      
295    United States
296            China
297    United States
298        Australia
299    United States
Name: country, Length: 300, dtype: object

In [11]:
# Select country and name  column 
data[["country","name"]]

country                                name
0            Italy                   Aarhus University
1            Spain    Arizona State University - Tempe
2    United States                   Auburn University
3           Canada      Australian National University
4            Italy  Autonomous University of Barcelona
..             ...                                 ...
295  United States           Xi'an Jiaotong University
296          China                     Yale University
297  United States                   Yonsei University
298      Australia                     York University
299  United States                 Zhejiang University

[300 rows x 2 columns]

6. Sort the data frame on the basis of values in a column.
- Each coulumn of dataFrame is treated as a pandas Series. The `.sort_values()` in DataFrames works similar to the `pandas.Series`.

In [12]:
data.head()

name   world_ranking region_ranking  \
0                   Aarhus University  #150 of 14,131   #44 of 2,785   
1    Arizona State University - Tempe   #61 of 14,131   #48 of 2,597   
2                   Auburn University             NaN            NaN   
3      Australian National University   #88 of 14,131       #5 of 59   
4  Autonomous University of Barcelona             NaN            NaN   

  country_ranking         region        country      city\state  \
0        #2 of 27         Europe          Italy          Veneto   
1    #45 of 2,496         Europe          Spain        Valencia   
2             NaN  North America  United States       Wisconsin   
3        #5 of 40  North America         Canada         Ontario   
4             NaN         Europe          Italy  Emilia-Romagna   

  acceptance_rate publication         website            phone_no  \
0             15%      89,633       www.au.dk       +45 8942 1111   
1             NaN      99,086     www.asu.edu          8552785080   
2             57%      36,231      auburn.edu          3348444000   
3             86%      97,754  www.anu.edu.au  +61 (0)2 6125 5111   
4             11%      74,922     www.uab.cat       +34 935812222   

                                             address  
0  Nordre Ringgade 1\n Aarhus, Central Denmark Re...  
1  University Drive and Mill Avenue\n Tempe, Ariz...  
2  Samford Hall\n Auburn, Alabama, 36849 \nUnited...  
3  Ellery Crescent, Acton\n Canberra, Australian ...  
4  Campus de Bellaterra, Edificio A\n Cerdanyola ...

In [13]:
# sort the data frame on basis of `publication` values
# by default the values will get sorted in ascending order
data.sort_values('publication')
# note: 'ascending = False' will sort the data frame in ascending order.

name   world_ranking region_ranking  \
197             University of Helsinki   #87 of 14,131   #17 of 2,785   
246            University of Rochester   #66 of 14,131   #50 of 2,597   
239                University of Padua  #151 of 14,131   #45 of 2,785   
164  University of California - Irvine   #56 of 14,131   #46 of 2,597   
130                Tel Aviv University  #119 of 14,131   #10 of 5,830   
..                                 ...             ...            ...   
1     Arizona State University - Tempe   #61 of 14,131   #48 of 2,597   
48                    Ghent University  #116 of 14,131   #30 of 2,785   
90              Normal Superior School  #166 of 14,131   #52 of 2,785   
117                Shandong University             NaN            NaN   
83    Nanyang Technological University             NaN            NaN   

    country_ranking         region         country              city\state  \
197        #1 of 36         Europe         Germany                  Saxony   
246    #47 of 2,496         Europe           Italy                Piedmont   
239        #4 of 88         Europe         Denmark                  Aarhus   
164    #43 of 2,496         Europe     Switzerland                   Basel   
130        #1 of 36  North America   United States          New York State   
..              ...            ...             ...                     ...   
1      #45 of 2,496         Europe           Spain                Valencia   
48         #2 of 63         Europe         Germany  North Rhine-Westphalia   
90        #1 of 224  North America   United States                 Florida   
117             NaN  North America   United States                    Ohio   
83              NaN         Europe  United Kingdom                 England   

    acceptance_rate publication          website              phone_no  \
197             35%     100,167  www.helsinki.fi         +358 (9) 1911   
246             NaN     100,238    rochester.edu            5852753221   
239             20%     100,320     www.unipd.it      +39 049 827 3131   
164             NaN     101,320          uci.edu            9498245011   
130             18%     101,800        tau.ac.il     +972 (3) 640 8111   
..              ...         ...              ...                   ...   
1               NaN      99,086      www.asu.edu            8552785080   
48              45%      99,103     www.ugent.be      +32 (9) 264 3011   
90              32%      99,156   www.ens.psl.eu  +33 (0)1 44 32 30 00   
117             21%      99,504       sdu.edu.cn   +86 (531) 883 64853   
83              56%      99,798   www.ntu.edu.sg         +65 6791 1744   

                                               address  
197  PO Box 33, Fabianinkatu 33\n Helsinki, Uusimaa...  
246  Wilson Blvd. - Wallis Hall\n Rochester, New Yo...  
239  Via VIII Febbraio, 2\n Padova, Veneto, 35122 \...  
164  501 Aldrich Hall\n Irvine, California, 92697 \...  
130  PO Box 39040\n Tel Aviv, Tel Aviv District, 69...  
..                                                 ...  
1    University Drive and Mill Avenue\n Tempe, Ariz...  
48   Sint-Pietersnieuwstraat 25\n Ghent, East Fland...  
90   45 Rue d'Ulm\n Paris, Ile-de-France, 75230 \nF...  
117  27 Shanda S Road, Licheng\n Jinan, Shandong, 2...  
83   50 Nanyang Avenue\n Singapore, 639798 \nSingapore  

[300 rows x 12 columns]

7. __Rank the dataframe__

In [15]:
# rank the data frame 'data' in descending order based on 'publication'
# 'method = min' assigns the minimum rank to highest equal value of `publication`
data['publication_rank'] = data['publication'].rank(ascending = 0,method = 'min')
data

name   world_ranking region_ranking  \
0                     Aarhus University  #150 of 14,131   #44 of 2,785   
1      Arizona State University - Tempe   #61 of 14,131   #48 of 2,597   
2                     Auburn University             NaN            NaN   
3        Australian National University   #88 of 14,131       #5 of 59   
4    Autonomous University of Barcelona             NaN            NaN   
..                                  ...             ...            ...   
295           Xi'an Jiaotong University             NaN            NaN   
296                     Yale University   #11 of 14,131    #9 of 2,597   
297                   Yonsei University             NaN            NaN   
298                     York University             NaN            NaN   
299                 Zhejiang University  #109 of 14,131    #7 of 5,830   

    country_ranking         region        country      city\state  \
0          #2 of 27         Europe          Italy          Veneto   
1      #45 of 2,496         Europe          Spain        Valencia   
2               NaN  North America  United States       Wisconsin   
3          #5 of 40  North America         Canada         Ontario   
4               NaN         Europe          Italy  Emilia-Romagna   
..              ...            ...            ...             ...   
295             NaN  North America  United States     Connecticut   
296     #9 of 2,496           Asia          China           Xi'an   
297             NaN  North America  United States      Washington   
298             NaN        Oceania      Australia        Canberra   
299       #3 of 960  North America  United States         Alabama   

    acceptance_rate publication           website             phone_no  \
0               15%      89,633         www.au.dk        +45 8942 1111   
1               NaN      99,086       www.asu.edu           8552785080   
2               57%      36,231        auburn.edu           3348444000   
3               86%      97,754    www.anu.edu.au   +61 (0)2 6125 5111   
4               11%      74,922       www.uab.cat        +34 935812222   
..              ...         ...               ...                  ...   
295              7%      96,097   www.xjtu.edu.cn    +86 (29) 266 8830   
296             NaN     198,095      www.yale.edu           2034324771   
297             56%      95,497  www.yonsei.ac.kr    +82 (2) 2123 2114   
298             33%      41,257          yorku.ca    +1 (416) 736 5002   
299             52%     176,136    www.zju.edu.cn  +86 (571) 8795 1020   

                                               address  publication_rank  
0    Nordre Ringgade 1\n Aarhus, Central Denmark Re...              29.0  
1    University Drive and Mill Avenue\n Tempe, Ariz...               5.0  
2    Samford Hall\n Auburn, Alabama, 36849 \nUnited...             188.0  
3    Ellery Crescent, Acton\n Canberra, Australian ...               8.0  
4    Campus de Bellaterra, Edificio A\n Cerdanyola ...              65.0  
..                                                 ...               ...  
295  28 Xianning Road\n Xi'an, Shaanxi, 710049 \nChina              12.0  
296  Woodbridge Hall\n New Haven, Connecticut, 0652...             216.0  
297  134 Sinchon-dong, Seodaemun-gu\n Seoul, Seoul,...              13.0  
298  4700 Keele Street\n Toronto, Ontario, M3J 1P3 ...             179.0  
299  38 Zheda Road, Xihu\n Hangzhou, Zhejiang, 3100...             229.0  

[300 rows x 13 columns]

From the above data frame , we can see that `BMI = 44.395875` is repeating thrice; thus the method = "min" will assign the minumum rank(=1) to all the three values of BMI. The rank `4` will be assigned to the second largest value of BMI and so on.Thus, there is no rank aqual to 2 and 3.

In [16]:
# method = 'dense' assigns some rank to all the same BMI values
data['publication_rank_dense'] = data['publication'].rank(method = 'dense')
data

name   world_ranking region_ranking  \
0                     Aarhus University  #150 of 14,131   #44 of 2,785   
1      Arizona State University - Tempe   #61 of 14,131   #48 of 2,597   
2                     Auburn University             NaN            NaN   
3        Australian National University   #88 of 14,131       #5 of 59   
4    Autonomous University of Barcelona             NaN            NaN   
..                                  ...             ...            ...   
295           Xi'an Jiaotong University             NaN            NaN   
296                     Yale University   #11 of 14,131    #9 of 2,597   
297                   Yonsei University             NaN            NaN   
298                     York University             NaN            NaN   
299                 Zhejiang University  #109 of 14,131    #7 of 5,830   

    country_ranking         region        country      city\state  \
0          #2 of 27         Europe          Italy          Veneto   
1      #45 of 2,496         Europe          Spain        Valencia   
2               NaN  North America  United States       Wisconsin   
3          #5 of 40  North America         Canada         Ontario   
4               NaN         Europe          Italy  Emilia-Romagna   
..              ...            ...            ...             ...   
295             NaN  North America  United States     Connecticut   
296     #9 of 2,496           Asia          China           Xi'an   
297             NaN  North America  United States      Washington   
298             NaN        Oceania      Australia        Canberra   
299       #3 of 960  North America  United States         Alabama   

    acceptance_rate publication           website             phone_no  \
0               15%      89,633         www.au.dk        +45 8942 1111   
1               NaN      99,086       www.asu.edu           8552785080   
2               57%      36,231        auburn.edu           3348444000   
3               86%      97,754    www.anu.edu.au   +61 (0)2 6125 5111   
4               11%      74,922       www.uab.cat        +34 935812222   
..              ...         ...               ...                  ...   
295              7%      96,097   www.xjtu.edu.cn    +86 (29) 266 8830   
296             NaN     198,095      www.yale.edu           2034324771   
297             56%      95,497  www.yonsei.ac.kr    +82 (2) 2123 2114   
298             33%      41,257          yorku.ca    +1 (416) 736 5002   
299             52%     176,136    www.zju.edu.cn  +86 (571) 8795 1020   

                                               address  publication_rank  \
0    Nordre Ringgade 1\n Aarhus, Central Denmark Re...              29.0   
1    University Drive and Mill Avenue\n Tempe, Ariz...               5.0   
2    Samford Hall\n Auburn, Alabama, 36849 \nUnited...             188.0   
3    Ellery Crescent, Acton\n Canberra, Australian ...               8.0   
4    Campus de Bellaterra, Edificio A\n Cerdanyola ...              65.0   
..                                                 ...               ...   
295  28 Xianning Road\n Xi'an, Shaanxi, 710049 \nChina              12.0   
296  Woodbridge Hall\n New Haven, Connecticut, 0652...             216.0   
297  134 Sinchon-dong, Seodaemun-gu\n Seoul, Seoul,...              13.0   
298  4700 Keele Street\n Toronto, Ontario, M3J 1P3 ...             179.0   
299  38 Zheda Road, Xihu\n Hangzhou, Zhejiang, 3100...             229.0   

     publication_rank_dense  
0                     272.0  
1                     296.0  
2                     113.0  
3                     293.0  
4                     236.0  
..                      ...  
295                   289.0  
296                    85.0  
297                   288.0  
298                   122.0  
299                    72.0  

[300 rows x 14 columns]

- Here , dense method assigns minimum rank (=1) to minimum value (=9.089335) of the BMI. Rank 2 will be assigned to BMI value greater than min=9.089335 and so on.Thus , no rank is skipped in the dense method. 
8. __To check for missing Values__
<!-- - We shall import a new dataset. -->

In [17]:
data.isnull().sum()

name                        0
world_ranking             101
region_ranking            101
country_ranking           101
region                      0
country                     0
city\state                  0
acceptance_rate            76
publication                 0
website                     6
phone_no                    4
address                     0
publication_rank            0
publication_rank_dense      0
dtype: int64

- The function `.isnull` check whether the data data is missing.The `sum()` sums the number of `True` values in a column. Th final output gives the number of missing values in each column.
- Here, we see there are 4 missing values in the `phone_no` column and one missing value in other columns.



In [18]:
data = {'A':[1,2,3,4,5], 'B':[1,0,1,1,0]}
df = pd.DataFrame(data=data)
df.C = df.A + df.B
df # Note that pandas doesnt allow to create new columns by using names.

C:\Users\user\AppData\Local\Temp/ipykernel_27144/3071025938.py:3: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  df.C = df.A + df.B


A  B
0  1  1
1  2  0
2  3  1
3  4  1
4  5  0

In [ ]:
df = pd.DataFrame({'a':[1,2,3],
                   'b':[4,5,6,],
                   'c':[7,8,9,]})

df_col_reversed = df.loc[:, ::-1] # reverse column order
df_row_reversed = df.loc[::-1, :] # reverse row order

5. __Reading Data from Diffent source values__
1.  Reading an `.xlsx` file.
- Pd can read excel but its formulas complicate things.
`pd.read_excel('example.xlsx')`
2. Read a Zip File.

```py
import zipfile
with zipfile.ZipFile('data.zip') as z:
    with z.open('exmaple.csv') as f:
        file = pd.read_csv(f) 
        print(file.head())
```
3. Read `.html` file
`df = pd.read_html("index.html",header=1,index_col=0)`
4. Read `.txt` file.
`pd.read_csv('example.txt',sep='\t')`
- You need to include if they are separators in the text files.
5. Read `.json` file
`pd.read_json('example.json',)`
- mostly used in IOT
6. Read `.xml` file
```xml
<?xml version="1.0" encoding="UTF-8"?>
<root>
	<bathrooms>
		<n35237 type="number">1.0</n35237>
		<n32238 type="number">3.0</n32238>
		<n44699 type="number">nan</n44699>
	</bathrooms>
	<price>
		<n35237 type="number">7020000.0</n35237>
		<n32238 type="number">10000000.0</n32238>
		<n44699 type="number">4128000.0</n44699>
	</price>
	<property_id>
		<n35237 type="number">35237.0</n35237>
		<n32238 type="number">32238.0</n32238>
		<n44699 type="number">44699.0</n44699>
	</property_id>
</root>
```

```py
import xml.etree.ElementTree as ET
import pandas as pd

xml_data = open('properties.xml', 'r').read()  # Read file
root = ET.XML(xml_data)  # Parse XML

data = []
cols = []
for i, child in enumerate(root):
    data.append([subchild.text for subchild in child])
    cols.append(child.tag)

df = pd.DataFrame(data).T  # Write in DF and transpose it
df.columns = cols  # Update column names
print(df)
```



### 1.2.8 Concatenation
- There various ways in which pandas combines data frames and then rearanges them.

## Table of Content
1. Concatenation
2. Merging and Joining 
3. Reshaping 
4. Pivot Table
5. Duplicate 
6. Map and Replace
7. Groupby in Pandas

1. __Concatenate__
- Concatenation is a process of joining objects along a particular axis, with set logic applied along the other axes, if any (a series doesnt have any other axes)
- The main parameters involved in pandas concatenation are - objects axis,handling of other axes, and keys.
- `pd.concat(objs,axis=0,join='outer',join_axes=None,ignore_index=False)`
objs - This is a sequence or mapping of series, Dataframe, or Panel Object.
- axis - `{0,1,...}`, default 0.This is the axis to concentrate along. 
- join `{'inner':'outer'}`, default 'outer'.How to handle indexes on other axis.Outer for union and inner for intersection
- ignore_index - boolean, default false.If True , do not use the index alues on the concatenation axis.The resulting axis will be labeled 0,...,n-1. 
- join_axes - This is the list of index objects. Specific indexes to use for the other(n-1) axes insteadof performing inner/outer set logic.
#### Concatenate Panda Series 


In [20]:
# Create series 
series_a = pd.Series([101,102,103,104,105,106,107,108,109])
series_b = pd.Series([111,112,113,114,115,116,117,118,119])

In [21]:
# concatenate series in pandas
pd.concat([series_a,series_b])

0    101
1    102
2    103
3    104
4    105
5    106
6    107
7    108
8    109
0    111
1    112
2    113
3    114
4    115
5    116
6    117
7    118
8    119
dtype: int64

### Add hierarchical index on Panda Series

In [23]:
pd.concat([series_a,series_b],keys=['a','b']) #----> this will indicate which part the series come from.

a  0    101
   1    102
   2    103
   3    104
   4    105
   5    106
   6    107
   7    108
   8    109
b  0    111
   1    112
   2    113
   3    114
   4    115
   5    116
   6    117
   7    118
   8    119
dtype: int64

#### Label the index

In [24]:
# 'names' parameter assigns labels to 'key' and 'indices'
pd.concat([series_a,series_b],keys=['a','b'],names=["Series","Row ID"])

Series  Row ID
a       0         101
        1         102
        2         103
        3         104
        4         105
        5         106
        6         107
        7         108
        8         109
b       0         111
        1         112
        2         113
        3         114
        4         115
        5         116
        6         117
        7         118
        8         119
dtype: int64

#### Concatenating data frames.
![Data Frame Concatenation](../../images/dfconcat.jpg)

In [5]:
df_1 = pd.DataFrame({
    'Name':['eby','edwin','charle'],
    'Company':['Ghent','dfdf','ArithmeticError'],
    'Salary':[20,40,56],
},index=[101,102,103])
df_2 = pd.DataFrame({
    'Name':['abby','edgar','earle'],
    'Company':['Ghnt','df','AritcError'],
    'Salary':[24,70,46],
},index=[101,102,103]) #specify indices

#### Using `.concat()`

In [6]:
pd.concat([df_1,df_2]) #this will just add the second at the bottom 

Name          Company  Salary
101     eby            Ghent      20
102   edwin             dfdf      40
103  charle  ArithmeticError      56
101    abby             Ghnt      24
102   edgar               df      70
103   earle       AritcError      46

##### Hozrizontal Concatenation
- They must have the same number of rows.

In [7]:
pd.concat([df_1,df_2],axis=1)

Name          Company  Salary   Name     Company  Salary
101     eby            Ghent      20   abby        Ghnt      24
102   edwin             dfdf      40  edgar          df      70
103  charle  ArithmeticError      56  earle  AritcError      46

##### Ignoring Indexes
- If `ignore_index = True`, Pandas does not use the index values along the concatenation axis. The resulting series/data frame will be labeled 0, ..., n - 1.

In [8]:
pd.concat([df_1,df_2],ignore_index=True)

Name          Company  Salary
0     eby            Ghent      20
1   edwin             dfdf      40
2  charle  ArithmeticError      56
3    abby             Ghnt      24
4   edgar               df      70
5   earle       AritcError      46

##### Concatenation with `.append()`

In [9]:
df_1.append(df_2)

Name          Company  Salary
101     eby            Ghent      20
102   edwin             dfdf      40
103  charle  ArithmeticError      56
101    abby             Ghnt      24
102   edgar               df      70
103   earle       AritcError      46

In [11]:
import numpy as np
ser1 = pd.Series(list('abcd'))
ser2 = pd.Series(np.arange(4))
df = pd.concat([ser1, ser2], axis=1)
df = pd.DataFrame({'col1': ser1, 'col2': ser2})
df

col1  col2
0    a     0
1    b     1
2    c     2
3    d     3

2. __Merging and Joining__

In [3]:
df_1 = pd.DataFrame({
    'Food_Product':['eby','edwin','charle'],
    'Brand':['Ghent','dfdf','ArithmeticError'],
    'Product_ID':[101,102,103],
})
df_2 = pd.DataFrame({
    'Food_Product':['ebdy','eddwin','cddddharle'],
    'Brand':['Ghent','ddddfdf','ArithdddddddmeticError'],
    'Product_ID':[101,103,102],
})

In [4]:
pd.merge(df_1,df_2,on='Product_ID') # This will merge based on the product ID. so every row with the same ID is merged

Food_Product_x          Brand_x  Product_ID Food_Product_y  \
0            eby            Ghent         101           ebdy   
1          edwin             dfdf         102     cddddharle   
2         charle  ArithmeticError         103         eddwin   

                  Brand_y  
0                   Ghent  
1  ArithdddddddmeticError  
2                 ddddfdf

- The output displays the common 'Product_ID' in the two data frames.
#### Merging Multiple Keys



In [5]:
pd.merge(df_1,df_2,on=['Product_ID','Food_Product'])

Empty DataFrame
Columns: [Food_Product, Brand_x, Product_ID, Brand_y]
Index: []

#### Merging on basis of index.
- This function helps to merge two dataframes, based on the index of both of them.
- This could be used to put two data frames side by side in their original forms.

In [6]:
pd.merge(df_1,df_2,right_index=True,left_index=True)

Food_Product_x          Brand_x  Product_ID_x Food_Product_y  \
0            eby            Ghent           101           ebdy   
1          edwin             dfdf           102         eddwin   
2         charle  ArithmeticError           103     cddddharle   

                  Brand_y  Product_ID_y  
0                   Ghent           101  
1                 ddddfdf           103  
2  ArithdddddddmeticError           102

- `merge()` allows you to perform natural join,lfet join ,right join,and full outer join in pandas.
#### Understanding the different types of merge
- Natural join: it keeps only rows that much from the data frames. (how='inner').
- Full outer join: To include all the rows of your data frame x and onlya those from y that match(how='left').
- Right outer join: To include all the rows of your data frame y and only those from x that match.(how='right').
![IMage](../../images/merge-types.jpg)
- __Inner Join__
    - It returns only the rows in which the left table has matching keys in the right table.

In [7]:
print(pd.merge(df_1,df_2,on='Product_ID',how='inner'))

  Food_Product_x          Brand_x  Product_ID Food_Product_y  \
0            eby            Ghent         101           ebdy   
1          edwin             dfdf         102     cddddharle   
2         charle  ArithmeticError         103         eddwin   

                  Brand_y  
0                   Ghent  
1  ArithdddddddmeticError  
2                 ddddfdf  


In [3]:
data1={'Physics': [77, 75, 100, 10, 59], 'Chemistry': [85, 70, 99, 30, 80]}
df1=pd.DataFrame(data=data1)
data2={'Student_ID': [0, 1, 2, 3, 4], 'Maths': [80, 90, 88, 25, 90]}
df2=pd.DataFrame(data=data2)
df3=pd.concat([df1, df2], join='inner', axis=0, ignore_index=True)
df3

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [4]:
data1={'Student_ID': [3, 4, 6, 8, 10], 'CGPA': [4.5, 3, 4.37, 3.5, 4]}
df1=pd.DataFrame(data=data1)
data2={'Student_ID': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 
'Maths': [4,52, 5, 2.5, 3, 3.9, 2.8, 4.75, 3.68, 5, 4.8]}
df2=pd.DataFrame(data=data2)
df3=pd.merge(df2, df1, on='Student_ID', how='left') 
df3

Student_ID  Maths  CGPA
0            0   4.00   NaN
1            1  52.00   NaN
2            2   5.00   NaN
3            3   2.50  4.50
4            4   3.00  3.00
5            5   3.90   NaN
6            6   2.80  4.37
7            7   4.75   NaN
8            8   3.68  3.50
9            9   5.00   NaN
10          10   4.80  4.00

In [5]:
header = pd.MultiIndex.from_product([['Before Course','After Course'],['Marks']])
d=([[82,95],[78,89]])
 
my_df = pd.DataFrame(d,
 index=['Alisa','Bobby'],
 columns=header)

print(my_df.stack(level=0).unstack(level=0))

              Marks      
              Alisa Bobby
After Course     95    89
Before Course    82    78


In [6]:
df = pd.DataFrame({"Gender": ["Male", "Male", "Female", "Female", "Female", "Female", "Male", "Male", "Female"],
 "Movie_Genre": ["Action", "Comedy", "Drama", "Action", "Comedy", "Drama", "Action", "Drama", "Action"],
 "Rating": [1, 5, 3, 2, 3, 4, 4, 5, 4]})
pd.pivot_table(data=df, index='Gender', values='Rating')

Rating
Gender        
Female    3.20
Male      3.75

In [7]:
pd.pivot_table(data=df, index='Gender', values='Rating', aggfunc='mean')

Rating
Gender        
Female    3.20
Male      3.75

In [8]:
pd.pivot_table(data=df, index='Gender', values='Rating')

Rating
Gender        
Female    3.20
Male      3.75

In [10]:
pd.pivot_table(data=df, index=['Gender', 'Movie_Genre'] , values='Rating')

Rating
Gender Movie_Genre        
Female Action          3.0
       Comedy          3.0
       Drama           3.5
Male   Action          2.5
       Comedy          5.0
       Drama           5.0

In [11]:
pd.pivot_table(data=df, index='Gender' , values='Rating', aggfunc='sum')

Rating
Gender        
Female      16
Male        15

In [12]:
# Get sum of both the genders ratings for each genre
pd.pivot_table(data=df, index=['Gender', 'Movie_Genre'] , values='Rating', aggfunc='sum')

Rating
Gender Movie_Genre        
Female Action            6
       Comedy            3
       Drama             7
Male   Action            5
       Comedy            5
       Drama             5

In [13]:
df_employee = [('John', 3400, 'Sydeny'),
 ('Robert', 3000, 'Chicago'),
 ('Aadi', 1600, 'New York'),
 ('Robert', 3000, 'Chicago'),
 ('Robert', 3000, 'Chicago'),
 ('Robert', 3000, 'Texas'),
 ('Aadi', 4000, 'London'),
 ('Sachin', 3000, 'Chicago')]
df_employee = pd.DataFrame(df_employee, columns=['Name', 'Salary', 'City'])
df_employee[df_employee.duplicated('Name')].shape

(4, 3)

In [23]:
data = {'Student1': {'name': 'Emma', 'age': '27', 'sex': 'Female'},
'Student2': {'name': 'Mike', 'age': '22', 'sex': 'Male'}}
df_students = pd.DataFrame(data=data)
df_students

Student1 Student2
name     Emma     Mike
age        27       22
sex    Female     Male

In [26]:
df_students['Student2'].replace('Mike', 'John', inplace=True)
df_students

Student1 Student2
name     Emma     John
age        27       22
sex    Female     Male